src/canisters/frontend/ic-certified-assets/src/system_context/mod.rs

```diff
@@ -0,0 +1,58 @@
+pub mod canister_env;
+
+use std::cell::{Ref, RefCell};
+
+use canister_env::CanisterEnv;
+use ic_cdk::api::time;
+
+/// Context that is available only inside canister runtime.
+///
+/// # Example
+///
+/// ```
+/// use ic_certified_assets::system_context::SystemContext;
+/// use ic_certified_assets::with_state_mut;
+/// use ic_certified_assets::types::CommitBatchArguments;
+/// use ic_cdk::api::{certified_data_set, trap};
+/// use ic_cdk::update;
+///
+/// #[update]
+/// pub fn commit_batch(arg: CommitBatchArguments) {
+///     let system_context = SystemContext::new();
+///
+///     with_state_mut(|s| {
+///         if let Err(msg) = s.commit_batch(arg, &system_context) {
+///             trap(&msg);
+///         }
+///         certified_data_set(s.root_hash());
+///     });
+/// }
+/// ```
+pub struct SystemContext {
+    /// Do not access this field directly, call [Self::get_canister_env] instead.
+    pub canister_env: RefCell<Option<CanisterEnv>>,
```

Comment:
```suggestion
    canister_env: RefCell<Option<CanisterEnv>>,
```
If you add a constructor function for tests only you can make the compiler enforce this